### PR TITLE
Adding writable and written books.

### PR DESCRIPTION
**Pull request by Sandertv**
## Introduction

This pull request adds Writable and written books to PocketMine. It adds API to modify books through plugins, as well as vanilla book writing behaviour.

### Relevant issues

Not relevant, feature addition.

### API changes

Newly registered Items:
* Writable Book (Book & Quill)
* Written Book

The WritableBook and WrittenBook classes contain API to modify the books through plugins, such as changing the author, modifying page text, adding new pages and swapping pages. The functions have documentation and should therefore speak for themselves.

### Behavioural changes

Players can now, once obtained, write in a Book & Quill and see their writing progress get saved. Books can also be signed now.

## Backwards compatibility

Not relevant.

## Follow-up

Actual tests.

## Tests

Functionality has been roughly tested, and most functions work as intended. There are however client issues that make a correct implementation impossible.

### CODE REVIEW

File: src/pocketmine/inventory/transaction/action/CreativeInventoryAction.php [hunk 1]
```diff
@@ -53,7 +53,7 @@ public function __construct(Item $sourceItem, Item $targetItem, int $actionType)
 	 */
 	public function isValid(Player $source) : bool{
 		return $source->isCreative(true) and
-			($this->actionType === self::TYPE_DELETE_ITEM or Item::getCreativeItemIndex($this->sourceItem) !== -1);
+			($this->actionType === self::TYPE_CREATE_ITEM or $this->actionType === self::TYPE_DELETE_ITEM or Item::getCreativeItemIndex($this->sourceItem) !== -1);
```
Comment (dktapps) [inline src/pocketmine/inventory/transaction/action/CreativeInventoryAction.php]: this will break creative-mode anti-cheat

File: src/pocketmine/network/mcpe/PlayerNetworkSessionAdapter.php [hunk 2]
```diff
@@ -227,6 +228,10 @@ public function handleResourcePackChunkRequest(ResourcePackChunkRequestPacket $p
 		return $this->player->handleResourcePackChunkRequest($packet);
 	}
 
+	public function handleBookEdit(BookEditPacket $packet): bool {
```
Comment (dktapps) [inline src/pocketmine/network/mcpe/PlayerNetworkSessionAdapter.php]: formatting

**Comment by SuperMaXAleX** [comment]
Oh yeeeah !

File: src/pocketmine/item/WrittenBook.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace pocketmine\item;
+
+class WrittenBook extends WritableBook{
+
+	public function __construct(int $meta = 0){
+		Item::__construct(self::WRITTEN_BOOK, $meta, "Written Book");
+	}
+
+	public function getMaxStackSize(): int {
```
Comment (dktapps) [inline src/pocketmine/item/WrittenBook.php]: formatting

File: src/pocketmine/Player.php [hunk 4]
```diff
@@ -2483,7 +2485,11 @@ public function handleMobEquipment(MobEquipmentPacket $packet) : bool{
 
 		$item = $this->inventory->getItem($packet->hotbarSlot);
 
-		if(!$item->equals($packet->item)){
+		if($item->getId() === Item::WRITABLE_BOOK){
+			if($packet->item->getId() === Item::WRITTEN_BOOK){
+				$this->inventory->setItem($packet->hotbarSlot, $packet->item);
```
Comment (dktapps) [inline src/pocketmine/Player.php]: this looks very hacky, why is this here?

**Review by dktapps — CHANGES_REQUESTED** [review]
please see my comments on the diff.

File: src/pocketmine/item/WritableBook.php [hunk 5]
```diff
@@ -20,11 +20,11 @@ public function __construct(int $meta = 0){
 		parent::__construct(self::WRITABLE_BOOK, $meta, "Book & Quill");
 	}
 
-	public function getMaxStackSize(): int {
+	public function getMaxStackSize() : int {
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: missed one here

Comment (Sandertv) [reply src/pocketmine/item/WritableBook.php]: Oops

**Comment by Sandertv** [comment]
I can't finish this implementation in a clean way due to a client issue. It sends the transaction packet required to swap a Book & Quill for a Written Book when signing the book, but this transaction gets sent after the MobEquipmentPacket, which expects the Book & Quill to already be a Written Book.

File: src/pocketmine/item/WritableBook.php [hunk 6]
```diff
@@ -175,7 +176,9 @@ public function setGeneration(int $generation) : void{
 	 * @return string
 	 */
 	public function getAuthor() : string{
-		$this->correctNBT();
+		if(!isset($this->getNamedTag()->author)){
```
Comment (TheDiamondYT1) [inline src/pocketmine/item/WritableBook.php]: Why not null?
```php
$sender->sendMessage($book->getAuthor() ?? TF::GREEN . "No author specified");
`

File: src/pocketmine/item/WritableBook.php [hunk 7]
```diff
@@ -207,9 +215,42 @@ public function getTitle() : string{
...
+	 * @param CompoundTag $namedTag
+	 *
+	 * @return bool
+	 */
+	private function movePagesAboveDownwards(int $id, CompoundTag $namedTag) : bool{
+		if(!isset($namedTag->pages)){
+			return false;
+		}
+		foreach($namedTag->pages as $key => $page){
+			if(!is_numeric($key) or $key <= $id){
```
Comment (Muqsit) [inline src/pocketmine/item/WritableBook.php]: `is_numeric` returns true for "floats", "negative numbers", floats and negative numbers. Since the keys are purely integers, maybe use `is_int`?

Comment (Sandertv) [reply src/pocketmine/item/WritableBook.php]: This function doesn't work as intended yet. is_int wouldn't work, as the $key would never be an integer. Instead, always a string containing an integer, so an int typecast would be required either way.

Comment (Muqsit) [reply src/pocketmine/item/WritableBook.php]: Ah, okay.

File: l [hunk 8]
```diff
@@ -0,0 +1,4 @@
+Item Book & Quill (386:0)x1 tags:
+0x0a000009050070616765730a0500000008090070686f746f6e616d6500000804007465787400000008090070686f746f6e616d6500000804007465787400000008090070686f746f6e616d6500000804007465787400000008090070686f746f6e616d6500000804007465787400000008090070686f746f6e616d6500000804007465787400000000
+Item Book & Quill (386:0)x1 tags:
+0x0a000009050070616765730a0700000008090070686f746f6e616d6500000804007465787400000008090070686f746f6e616d6500000804007465787400000008040074657874000008090070686f746f6e616d6500000008040074657874000008090070686f746f6e616d6500000008040074657874000008090070686f746f6e616d6500000008040074657874000008090070686f746f6e616d6500000008040074657874000008090070686f746f6e616d6500000000
```
Comment (dktapps) [inline l]: i don't think you meant to commit this

Comment (Sandertv) [reply l]: Lmao no 😄

Comment (bontaki2) [reply l]: my name is jsonwynn10 and I like every single fucking comment on github

File: src/pocketmine/inventory/transaction/action/CreativeInventoryAction.php [hunk 9]
```diff
@@ -33,7 +33,7 @@ class CreativeInventoryAction extends InventoryAction{
 	 */
 	const TYPE_DELETE_ITEM = 0;
 	/**
-	 * Player took an item from the creative window.
+	 * Player took an item from the creative window or signs a book.
```
Comment (dktapps) [inline src/pocketmine/inventory/transaction/action/CreativeInventoryAction.php]: this can be removed

Comment (Sandertv) [reply src/pocketmine/inventory/transaction/action/CreativeInventoryAction.php]: It does still get sent for creative mode book signing though.

Comment (dktapps) [reply src/pocketmine/inventory/transaction/action/CreativeInventoryAction.php]: yes, but you're not using that, and it's likely that that is a bug and not intended (mismatch in survival)

File: src/pocketmine/Player.php [hunk 10]
```diff
@@ -3060,40 +3060,40 @@ public function handleItemFrameDropItem(ItemFrameDropItemPacket $packet) : bool{
...
 			default:
 				return false;
 		}
-		$this->getInventory()->setItem($packet->inventorySlot - 9, $book);
+		$this->getServer()->getPluginManager()->callEvent($event = new PlayerEditBookEvent($this, $oldBook, $newBook, $packet->type));
+		if($event->isCancelled()){
+			return true;
+		}
+		$newBook = $event->getNewBook();
+		if($packet->type === 4){
```
Comment (dktapps) [inline src/pocketmine/Player.php]: why is this hardcoded literal 4? and why is it not in the case statement?

File: src/pocketmine/item/WritableBook.php [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+*/
+
+declare(strict_types = 1);
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: nit: the code style prefers no spaces here.

File: src/pocketmine/event/player/PlayerEditBookEvent.php [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+*/
+
+declare(strict_types = 1);
```
Comment (dktapps) [inline src/pocketmine/event/player/PlayerEditBookEvent.php]: same here.

File: src/pocketmine/item/WrittenBook.php [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+*/
+
+declare(strict_types = 1);
```
Comment (dktapps) [inline src/pocketmine/item/WrittenBook.php]: and here.

File: src/pocketmine/item/WritableBook.php (hunk 11 above)
Comment (Sandertv) [reply src/pocketmine/item/WritableBook.php]: Oops, my bad.

File: src/pocketmine/item/WritableBook.php [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+use pocketmine\nbt\tag\IntTag;
+use pocketmine\nbt\tag\ListTag;
+use pocketmine\nbt\tag\StringTag;
+
+class WritableBook extends Item{
+
+	const GENERATION_ORIGINAL = 0;
+	const GENERATION_COPY = 1;
+	const GENERATION_COPY_OF_COPY = 2;
+	const GENERATION_TATTERED = 3;
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: shouldn't these be in `WrittenBook` and not here?

File: src/pocketmine/item/WritableBook.php [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+		return true;
+	}
+
+	/**
+	 * Returns the generation of the book.
+	 * Generations higher than 1 can not be copied.
+	 *
+	 * @return int
+	 */
+	public function getGeneration() : int{
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: Only written books may have a generation. See the Minecraft Wiki.

File: src/pocketmine/item/WritableBook.php [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+		$this->setNamedTag($namedTag);
+	}
+
+	/**
+	 * Returns the author of this book.
+	 * This is not a reliable way to get the real author of the book. It can be changed in-game.
+	 *
+	 * @return string
+	 */
+	public function getAuthor() : string{
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: same as above

File: src/pocketmine/item/WritableBook.php [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+		}
+		$this->setNamedTag($namedTag);
+	}
+
+	/**
+	 * Returns the title of this book.
+	 *
+	 * @return string
+	 */
+	public function getTitle() : string{
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: same as above

File: src/pocketmine/item/WritableBook.php [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+		}
+		return $this->getNamedTag()->author->getValue();
+	}
+
+	/**
+	 * Sets the author of this book.
+	 *
+	 * @param string $authorName
+	 */
+	public function setAuthor(string $authorName) : void{
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: ^

File: src/pocketmine/item/WritableBook.php [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+		}
+		return $this->getNamedTag()->title->getValue();
+	}
+
+	/**
+	 * Sets the author of this book.
+	 *
+	 * @param string $title
+	 */
+	public function setTitle(string $title) : void{
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: ^

File: src/pocketmine/item/WritableBook.php [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+		}
+		return $this->getNamedTag()->generation->getValue();
+	}
+
+	/**
+	 * Sets the generation of a book.
+	 *
+	 * @param int $generation
+	 */
+	public function setGeneration(int $generation) : void{
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: ^

**Review by dktapps — DISMISSED** [review]
According to the Minecraft Wiki, only Written Books have generation, title and author properties. These methods and constants are therefore incorrectly placed.

Reference material:
https://minecraft.gamepedia.com/Book_and_Quill
https://minecraft.gamepedia.com/Written_Book

File: src/pocketmine/item/WrittenBook.php [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+namespace pocketmine\item;
+
+class WrittenBook extends WritableBook{
+
+	public function __construct(int $meta = 0){
+		Item::__construct(self::WRITTEN_BOOK, $meta, "Written Book");
+	}
+
+	public function getMaxStackSize() : int{
+		return 16;
```
Comment (dktapps) [inline src/pocketmine/item/WrittenBook.php]: ?????

File: src/pocketmine/item/WritableBook.php (hunk 14 above)
Comment (dktapps) [reply src/pocketmine/item/WritableBook.php]: bUmP

File: src/pocketmine/item/WritableBook.php [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+			}
+			$namedTag->pages->{$key + $type} = $page;
+		}
+		return true;
+	}
+
+	/**
+	 * Returns an array containing all pages of this book.
+	 *
+	 * @return array
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: what is this an array _of_?

File: src/pocketmine/item/WrittenBook.php [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+			$namedTag->generation->setValue($generation);
+		}else{
+			$namedTag->generation = new IntTag("generation", $generation);
+		}
+		$this->setNamedTag($namedTag);
+	}
+
+	/**
+	 * Returns the author of this book.
+	 * This is not a reliable way to get the real author of the book. It can be changed in-game.
```
Comment (dktapps) [inline src/pocketmine/item/WrittenBook.php]: is this still applicable?

File: src/pocketmine/item/WritableBook.php [hunk 24]
```diff
@@ -0,0 +1,233 @@
...
+
+		return $created;
+	}
+
+	/**
+	 * Adds a new page with the given text. (if given)
+	 *
+	 * @param int    $pageId
+	 *
+	 * @return int page number
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: why are you returning the same thing the code just gave you?

**Comment by dktapps** [comment]
@Sandertv sorry, my network connection is terrible, I seem to only be able to comment on github. I'll disable interaction limits.

File: src/pocketmine/item/WritableBook.php (hunk 14 above)
Comment (Sandertv) [reply src/pocketmine/item/WritableBook.php]: Fixing next commit.

File: src/pocketmine/item/WrittenBook.php (hunk 23 above)
Comment (dktapps) [reply src/pocketmine/item/WrittenBook.php]: I'm not sure I understand you... written books cannot be modified after being signed.

Comment (Sandertv) [reply src/pocketmine/item/WrittenBook.php]: During the process of signing a book, the player can give the book a name, and a PE only feature is that you can enter any name you want as the signer of the book.

Comment (dktapps) [reply src/pocketmine/item/WrittenBook.php]: you're missing my point... a _written_ book cannot be modified. Therefore this is a "reliable" method in terms of "it won't change", however you haven't specified that "this can be anything, not just a player name". The documentation is too ambiguous.

Comment (Sandertv) [reply src/pocketmine/item/WrittenBook.php]: Got it, see latest commit.

File: src/pocketmine/item/WritableBook.php [hunk 25]
```diff
@@ -0,0 +1,228 @@
...
+	 * @param int $pageId
+	 *
+	 * @return bool
+	 */
+	public function pageExists(int $pageId) : bool{
+		return isset($this->getNamedTag()->pages->{$pageId});
+	}
+
+	/**
+	 * Returns a string containing the content of a page, or an empty string if the page doesn't exist.
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: this documentation is inaccurate (null, not empty string)

File: src/pocketmine/item/WritableBook.php [hunk 26]
```diff
@@ -0,0 +1,228 @@
...
+
+		for($id = 0; $id <= $pageId; $id++) {
+			if(!$this->pageExists($id)) {
+				$namedTag->pages->{$id} = new CompoundTag("", [
+					new StringTag("text", ""),
+					new StringTag("photoname", "")
+				]);
+			}
+		}
+
```
Comment (dktapps) [inline src/pocketmine/item/WritableBook.php]: nitpick: some formatting issues in this function

Comment (Sandertv) [reply src/pocketmine/item/WritableBook.php]: Fixed those just now.

**Comment by dktapps** [comment]
Resolved merge conflicts, please pull before making any futher changes.

**Review by dktapps — COMMENTED** [review]
you forgot encode

**Review by dktapps — APPROVED** [review]
